flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/types/SlotProfile.java

```diff
@@ -203,68 +206,104 @@ public LocalityAwareRequirementsToSlotMatcher(@Nonnull Collection<TaskManagerLoc
+					if (hostLocalWeigh != null) {
 						if (additionalRequirementsFilter.test(candidate)) {
-							// We remember the candidate, but still continue because there might still be a candidate
-							// that is local to the desired task manager.
-							matchByHostName = candidate;
+							// we found a match with host locality.
+							candidateMatchedResults.put(candidate, new CandidateMatchedResult(0, hostLocalWeigh));
 						} else {
 							// next candidate because this failed on the additional requirements.
 							continue;
 						}
 					}
+				}
 
-					// this if checks if the candidate at least fulfils the resource requirements, and is only required
-					// if we did not yet find a valid candidate with better locality.
-					if (matchByAdditionalRequirements == null
-						&& additionalRequirementsFilter.test(candidate)) {
-						// Again, we remember but continue in hope for a candidate with better locality.
-						matchByAdditionalRequirements = candidate;
-					}
+				// this if checks if the candidate at least fulfils the resource requirements, and is only required
+				// if we did not yet find a valid candidate with better locality.
+				if (candidateMatchedResults.isEmpty()
+					&& matchByAdditionalRequirements == null
+					&& additionalRequirementsFilter.test(candidate)) {
+					// Again, we remember but continue in hope for a candidate with better locality.
+					matchByAdditionalRequirements = candidate;
 				}
 			}
 
+			// find the best matched one.
+			Map.Entry<IN, CandidateMatchedResult> theBestOne = candidateMatchedResults.entrySet()
+					.stream()
+					.sorted(Comparator.comparingInt((Map.Entry<IN, CandidateMatchedResult> matchResult)
+						-> matchResult.getValue().getScore()).reversed())
+					.findFirst()
+					.orElse(null);
+
 			// at the end of the iteration, we return the candidate with best possible locality or null.
-			if (matchByHostName != null) {
-				return resultProducer.apply(matchByHostName, Locality.HOST_LOCAL);
+			if (theBestOne != null) {
+				return resultProducer.apply(theBestOne.getKey(), theBestOne.getValue().getLocality());
 			} else if (matchByAdditionalRequirements != null) {
 				return resultProducer.apply(matchByAdditionalRequirements, Locality.NON_LOCAL);
 			} else {
 				return null;
 			}
 		}
+
+		/**
+		 * Helper class to record the match result.
+		 */
+		private class CandidateMatchedResult {
+
+			private int localCount;
+
+			private int hostLocalCount;
+
+			public CandidateMatchedResult(int localCount, int hostLocalCount) {
+				this.localCount = localCount;
+				this.hostLocalCount = hostLocalCount;
+			}
+
+			// evaluate the match score
+			public int getScore() {
+				return localCount * 10 + hostLocalCount * 1;
+			}
+
+			// get the highest locality.
+			public Locality getLocality() {
+				return localCount > 0 ? Locality.LOCAL : Locality.HOST_LOCAL;
```

Comment:
What if also `hostLocal == 0`? We do still return `Locality.HOST_LOCAL`?

Reply:
This is won't happen in our case, but you are right this is a logical bug, will change.